controllers/condition/controller.go

```diff
@@ -0,0 +1,140 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package condition
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/controllers/common"
+)
+
+// Reconciler for common chaos
+type Reconciler struct {
+	// Object is used to mark the target type of this Reconciler
+	Object runtime.Object
+
+	// Client is used to operate on the Kubernetes cluster
+	client.Client
+
+	Recorder record.EventRecorder
+
+	Log logr.Logger
+}
+
+type StatusAndReason struct {
+	Status corev1.ConditionStatus
+	Reason string
+}
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	obj := r.Object.DeepCopyObject().(common.InnerObjectWithSelector)
+
+	if err := r.Client.Get(context.TODO(), req.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.Log.Info("chaos not found")
+		} else {
+			// TODO: handle this error
+			r.Log.Error(err, "unable to get chaos")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	conditionMap := make(map[v1alpha1.ChaosConditionType]StatusAndReason)
+	for _, c := range obj.GetStatus().Conditions {
+		conditionMap[c.Type] = StatusAndReason{
+			Status: c.Status,
+			Reason: c.Reason,
+		}
+	}
+
+	newConditionMap := make(map[v1alpha1.ChaosConditionType]StatusAndReason)
+	if obj.GetStatus().Experiment.Records != nil {
+		newConditionMap[v1alpha1.ConditionSelected] = StatusAndReason{
+			Status: corev1.ConditionTrue,
+		}
+	} else {
+		newConditionMap[v1alpha1.ConditionSelected] = StatusAndReason{
+			Status: corev1.ConditionFalse,
+		}
+	}
+
+	allInjected := corev1.ConditionTrue
+	allRecovered := corev1.ConditionTrue
+	for _, record := range obj.GetStatus().Experiment.Records {
+		if record.Phase != v1alpha1.NotInjected {
+			allRecovered = corev1.ConditionFalse
+		}
+
+		if record.Phase != v1alpha1.Injected {
+			allInjected = corev1.ConditionFalse
+		}
+	}
+	newConditionMap[v1alpha1.ConditionAllInjected] = StatusAndReason{
+		Status: allInjected,
+	}
+	newConditionMap[v1alpha1.ConditionAllRecovered] = StatusAndReason{
+		Status: allRecovered,
+	}
+
+	if obj.IsPaused() {
+		newConditionMap[v1alpha1.ConditionPaused] = StatusAndReason{
+			Status: corev1.ConditionTrue,
+		}
+	} else {
+		newConditionMap[v1alpha1.ConditionPaused] = StatusAndReason{
+			Status: corev1.ConditionFalse,
+		}
+	}
+
+	if !reflect.DeepEqual(newConditionMap, conditionMap) {
+		conditions := make([]v1alpha1.ChaosCondition, 0, 5)
+		for k, v := range newConditionMap {
+			conditions = append(conditions, v1alpha1.ChaosCondition{
+				Type:   k,
+				Status: v.Status,
+				Reason: v.Reason,
```

Comment:
How to fill out the `reason`?

We did have `FailedMessage` to store the error message while applied failed or recover failed. For now, it seems that we lost this message?

Reply:
There is only one reason in these condition, and they are in the `if` statement :yum:.

For example, in the conditions of a `Pod`, the reason is `ContainersNotReady` but not `ImagePullBackOff`. If we need to find a place to write the `FailedMessage`, a `recordStatuses` or adding status for every record is better.

Now, the error messages are only sent through events. I could add a status for every record later :smile_cat: .https://github.com/chaos-mesh/chaos-mesh/issues/1833 an issue has been recorded.

Reply:
IMO, I think `reason` field is optional, rather than required as we defined [here](https://github.com/chaos-mesh/chaos-mesh/pull/1832/files#diff-1b59fa0b31ead22983a731ae3671d88df11dcfb6d0ba7bb84cdaeca7d50f9c4dR92).

It would be empty if the condition status is `true`. If the condition status is `false`, the `reason` should be sth like one of the preset enums, like Pod Conditions [consts](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/status/generate.go#L33).

The reason field should tell the general failed reason, and just like the `Pod` resource did, the details, as you mentioned, would be better in sth like `recordStatuses`.

Another question is why we need a separate condition controller? We did apply action at [here](https://github.com/chaos-mesh/chaos-mesh/blob/nirvana/controllers/common/controller.go#L123), and now we just send events while it fails, and why not update its conditions accordingly. Just like kubelet did [syncPod](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1490) and it sends [event](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L180) if ImagePullBackOff, [updates conditions](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1711) inside this sync loop as well.